File: library/src/main/java/com/google/maps/android/collections/MapObjectManager.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.maps.android.collections;

import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;

import com.google.android.gms.maps.GoogleMap;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

/**
 * Abstract base implementation for map object collection manager classes.
 * <p/>
 * Keeps track of collections of objects on the map. Delegates all object-related events to each
 * collection's individually managed listeners.
 * <p/>
 * All object operations (adds and removes) should occur via its collection class. That is, don't
 * add an object via a collection, then remove it via Object.remove()
 */
abstract class MapObjectManager<O, C extends MapObjectManager.Collection> {
    protected final GoogleMap mMap;

    private final Map<String, C> mNamedCollections = new HashMap<>();
    protected final Map<O, C> mAllObjects = new HashMap<>();

    public MapObjectManager(@NonNull GoogleMap map) {
        mMap = map;
        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                setListenersOnUiThread();
            }
        });
    }

    abstract void setListenersOnUiThread();

    public abstract C newCollection();

    /**
     * Create a new named collection, which can later be looked up by {@link #getCollection(String)}
     *
     * @param id a unique id for this collection.
     */
    public C newCollection(String id) {
        if (mNamedCollections.get(id) != null) {
            throw new IllegalArgumentException("collection id is not unique: " + id);
        }
        C collection = newCollection();
        mNamedCollections.put(id, collection);
        return collection;
    }

    /**
     * Gets a named collection that was created by {@link #newCollection(String)}
     *
     * @param id the unique id for this collection.
     */
    public C getCollection(String id) {
        return mNamedCollections.get(id);
    }

    /**
     * Removes an object from its collection.
     *
     * @param object the object to remove.
     * @return true if the object was removed.
     */
    public boolean remove(O object) {
        C collection = mAllObjects.get(object);
        return collection != null && collection.remove(object);
    }

    protected abstract void removeObjectFromMap(O object);

    public class Collection {
        private final Set<O> mObjects = new LinkedHashSet<>();

        public Collection() {
        }

        protected void add(O object) {
            mObjects.add(object);
            mAllObjects.put(object, (C) this);
        }

        protected boolean remove(O object) {
            if (mObjects.remove(object)) {
                mAllObjects.remove(object);
                removeObjectFromMap(object);
                return true;
            }
            return false;
        }

        public void clear() {
            for (O object : mObjects) {
                removeObjectFromMap(object);
                mAllObjects.remove(object);
            }
            mObjects.clear();
        }

        protected java.util.Collection<O> getObjects() {
            return Collections.unmodifiableCollection(mObjects);
        }
    }
}

File: library/src/androidMain/kotlin/com/google/maps/android/collections/MapObjectManager.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.maps.android.collections

import android.os.Handler
import android.os.Looper
import com.google.android.gms.maps.GoogleMap
import kotlin.collections.Collection as KotlinCollection

/**
 * Abstract base implementation for map object collection manager classes.
 *
 * Keeps track of collections of objects on the map. Delegates all object-related events to each
 * collection's individually managed listeners.
 *
 * All object operations (adds and removes) should occur via its collection class. That is, don't
 * add an object via a collection, then remove it via Object.remove()
 */
abstract class MapObjectManager<O : Any, C : MapObjectManager<O, C>.Collection>(
    @JvmField
    protected val mMap: GoogleMap,
) {
    private val mNamedCollections: MutableMap<String, C> = mutableMapOf()

    @JvmField
    protected val mAllObjects: MutableMap<O, C> = mutableMapOf()

    init {
        Handler(Looper.getMainLooper()).post {
            setListenersOnUiThread()
        }
    }

    internal abstract fun setListenersOnUiThread()

    abstract fun newCollection(): C

    /**
     * Create a new named collection, which can later be looked up by [getCollection]
     *
     * @param id a unique id for this collection.
     */
    open fun newCollection(id: String): C {
        require(mNamedCollections[id] == null) { "collection id is not unique: $id" }
        val collection = newCollection()
        mNamedCollections[id] = collection
        return collection
    }

    /**
     * Gets a named collection that was created by [newCollection]
     *
     * @param id the unique id for this collection.
     */
    open fun getCollection(id: String): C? = mNamedCollections[id]

    /**
     * Removes an object from its collection.
     *
     * @param mapObject the object to remove.
     * @return true if the object was removed.
     */
    open fun remove(mapObject: O?): Boolean =
        mapObject != null && mAllObjects[mapObject]?.remove(mapObject) == true

    protected abstract fun removeObjectFromMap(mapObject: O)

    protected open fun setVisible(mapObject: O, visible: Boolean) {}

    open inner class Collection {
        private val mObjects: MutableSet<O> = mutableSetOf()

        // Safe unchecked cast: this inner collection is an instance of subclass C.
        @Suppress("UNCHECKED_CAST")
        protected open fun add(mapObject: O) {
            mObjects.add(mapObject)
            mAllObjects[mapObject] = this@Collection as C
        }

        protected open fun checkAndAdd(mapObject: O?, typeName: String): O =
            checkNotNull(mapObject) { "Failed to add $typeName to GoogleMap" }.also { add(it) }

        open fun showAll() {
            for (mapObject in mObjects) {
                setVisible(mapObject, true)
            }
        }

        open fun hideAll() {
            for (mapObject in mObjects) {
                setVisible(mapObject, false)
            }
        }

        protected open fun <T> addAll(opts: KotlinCollection<T>, adder: (T) -> O) {
            for (opt in opts) {
                adder(opt)
            }
        }

        protected open fun <T> addAll(opts: KotlinCollection<T>, defaultVisible: Boolean, adder: (T) -> O) {
            for (opt in opts) {
                val obj = adder(opt)
                setVisible(obj, defaultVisible)
            }
        }

        open fun remove(mapObject: O?): Boolean {
            if (mapObject == null) return false
            if (mObjects.remove(mapObject)) {
                mAllObjects.remove(mapObject)
                removeObjectFromMap(mapObject)
                return true
            }
            return false
        }

        open fun clear() {
            for (mapObject in mObjects) {
                removeObjectFromMap(mapObject)
                mAllObjects.remove(mapObject)
            }
            mObjects.clear()
        }

        protected open fun getObjects(): KotlinCollection<O> = mObjects
    }
}
